File: 2020/day1/part1.cpp
#include <bits/stdc++.h>

#define MAX_LENGTH 128000

// Part1
// solution in O(n) time-complexity & O(n) memory
uint64_t solve(std::vector<int> vec){
    const int target = 2020;
    std::vector<int> mmap(target+1, 0);
    uint64_t result = 1;

    //O(n)
    for(auto &v: vec)
        mmap[v] += 1;

    //O(n)
    for(int i=0; i<=target/2; i++)
        result *= std::pow(i*(target-i), std::min(mmap[i], mmap[target-i]));

    return result;
}


int main(int argc, char *argv[]){

    std::ifstream fin;
    std::vector<int> entries;

    fin.open("input.txt");

    while(fin) {

        std::string a; 
        std::getline(fin, a);
        
        try {
            entries.push_back(std::stoi(a));
        } catch(const std::exception& e) {
            std::cout << e.what() << std::endl;
        }

    }

    // Print the solution (result)
    std::cout << solve(entries) << std::endl;

    return 0;
}

File: 2020/13-Shuttle-Search/part1.py
def solve(time, ids):
    optimum =  min(ids, key= lambda x: x-(time%x))
    return (optimum-(time%optimum))*optimum


if __name__ == "__main__":
    with open('input.txt') as fp:
        lines = fp.readlines()
        time_stamp = int(lines[0])
        bus_ids = list()
        for line in lines[1:]:
            for bus_id in line[:-1].split(','):
                if bus_id != 'x':
                    bus_ids.append(int(bus_id))
    result = solve(time_stamp, bus_ids)
    print(result)
